File: Python/Heap_Sort.py
# Heapsort es un algoritmo de ordenamiento basado en un árbol binario llamado "heap".
# La idea es construir un heap máximo (max-heap) y luego extraer el mayor elemento 
# (la raíz) para colocarlo al final del arreglo, repitiendo el proceso hasta ordenar todo.

def heapify(arr, n, i):
    # Función para asegurar que el subárbol con raíz en el índice i
    # cumpla con la propiedad de heap máximo.
    # arr: lista a ordenar
    # n: tamaño del heap
    # i: índice de la raíz del subárbol
    
    largest = i        # Inicialmente, asumimos que la raíz es el mayor
    left = 2 * i + 1   # Índice del hijo izquierdo
    right = 2 * i + 2  # Índice del hijo derecho

    # Si el hijo izquierdo existe y es mayor que la raíz
    if left < n and arr[left] > arr[largest]:
        largest = left

    # Si el hijo derecho existe y es mayor que el mayor actual
    if right < n and arr[right] > arr[largest]:
        largest = right

    # Si el mayor no es la raíz, intercambiamos y seguimos ajustando recursivamente
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Intercambio
        heapify(arr, n, largest)  # Ajustar el subárbol afectado


def heapsort(arr):
    # Función principal de Heapsort.
    # Ordena la lista arr en orden ascendente.
    
    n = len(arr)

    # Paso 1: Construir un heap máximo
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Paso 2: Extraer elementos uno por uno del heap
    for i in range(n - 1, 0, -1):
        # Mover la raíz (máximo) al final
        arr[0], arr[i] = arr[i], arr[0]
        
        # Llamar heapify en el heap reducido
        heapify(arr, i, 0)


# Ejemplo de uso
if __name__ == "__main__":
    lista = [12, 11, 13, 5, 6, 7]
    print("Antes de ordenar los elementos del array son: ", lista)
    heapsort(lista)
    print("\nDespués de ordenar el arreglo: ", lista)

File: C++/Heap_sort.cpp
#include <iostream>
using namespace std;

// Función para ajustar un subárbol al heap máximo
void heapify(int arr[], int n, int i) {
    int largest = i;          // Inicialmente, asumimos que la raíz es el mayor
    int left = 2 * i + 1;     // Hijo izquierdo
    int right = 2 * i + 2;    // Hijo derecho

    // Si el hijo izquierdo es mayor que la raíz
    if (left < n && arr[left] > arr[largest])
        largest = left;

    // Si el hijo derecho es mayor que el mayor actual
    if (right < n && arr[right] > arr[largest])
        largest = right;

    // Si el mayor no es la raíz
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

// Implementación de Heapsort
void heapsort(int arr[], int n) {
    // Construir heap máximo
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);

    // Extraer elementos uno a uno
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]); // Mover raíz al final
        heapify(arr, i, 0);   // Heapify en el heap reducido
    }
}

// Programa principal
int main() {
    int arr[] = {12, 11, 13, 5, 6, 7};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Antes de ordenar: ";
    for (int i = 0; i < n; i++) cout << arr[i] << " ";
    cout << endl;

    heapsort(arr, n);

    cout << "Después de ordenar: ";
    for (int i = 0; i < n; i++) cout << arr[i] << " ";
    cout << endl;

    return 0;
}